File: src/main/java/net/jgp/books/spark/ch12/lab951_csv_with_embdedded_json_auto/CsvWithEmbdeddedJsonAutomaticJsonifierApp.java
package net.jgp.books.spark.ch12.lab951_csv_with_embdedded_json_auto;

import static org.apache.spark.sql.functions.concat;
import static org.apache.spark.sql.functions.explode;
import static org.apache.spark.sql.functions.lit;

import java.io.Serializable;
import java.util.Arrays;
import java.util.List;

import org.apache.spark.api.java.function.MapFunction;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Encoders;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.StructField;

/**
 * Ingesting a CSV with embedded JSON.
 * 
 * @author jgp
 */
public class CsvWithEmbdeddedJsonAutomaticJsonifierApp
    implements Serializable {
  private static final long serialVersionUID = 19713L;

  /**
   * Turns a Row into JSON.
   * 
   * @author jgp
   */
  private final class Jsonifier
      implements MapFunction<Row, String> {
    private static final long serialVersionUID = 19714L;

    private StructField[] fields = null;
    private List<String> jsonColumns = null;

    public Jsonifier(String... jsonColumns) {
      this.fields = null;
      this.jsonColumns = Arrays.asList(jsonColumns);
    }

    @Override
    public String call(Row r) throws Exception {
      if (fields == null) {
        fields = r.schema().fields();
      }
      StringBuffer sb = new StringBuffer();
      sb.append('{');
      int fieldIndex = -1;
      boolean isJsonColumn;
      for (StructField f : fields) {
        isJsonColumn = false;
        fieldIndex++;
        if (fieldIndex > 0) {
          sb.append(',');
        }
        if (this.jsonColumns.contains(f.name())) {
          isJsonColumn = true;
        }
        if (!isJsonColumn) {
          sb.append('"');
          sb.append(f.name());
          sb.append("\": ");
        }
        String type = f.dataType().toString();
        switch (type) {
          case "IntegerType":
            sb.append(r.getInt(fieldIndex));
            break;
          case "LongType":
            sb.append(r.getLong(fieldIndex));
            break;
          case "DoubleType":
            sb.append(r.getDouble(fieldIndex));
            break;
          case "FloatType":
            sb.append(r.getFloat(fieldIndex));
            break;
          case "ShortType":
            sb.append(r.getShort(fieldIndex));
            break;

          default:
            if (isJsonColumn) {
              // JSON field
              String s = r.getString(fieldIndex);
              if (s != null) {
                s = s.trim();
                if (s.charAt(0) == '{') {
                  s = s.substring(1, s.length() - 1);
                }
              }
              sb.append(s);
            } else {
              sb.append('"');
              sb.append(r.getString(fieldIndex));
              sb.append('"');
            }
            break;
        }
      }
      sb.append('}');
      return sb.toString();
    }
  }

  /**
   * main() is your entry point to the application.
   * 
   * @param args
   */
  public static void main(String[] args) {
    CsvWithEmbdeddedJsonAutomaticJsonifierApp app =
        new CsvWithEmbdeddedJsonAutomaticJsonifierApp();
    app.start();
  }

  /**
   * The processing code.
   */
  private void start() {
    // Creates a session on a local master
    SparkSession spark = SparkSession.builder()
        .appName("Processing of invoices")
        .master("local[*]")
        .getOrCreate();

    Dataset<Row> df = spark
        .read()
        .option("header", true)
        .option("delimiter", "|")
        .option("inferSchema", true)
        .csv("data/misc/csv_with_embedded_json2.csv");
    df.show(5, false);
    df.printSchema();

    Dataset<String> ds = df.map(
        new Jsonifier("emp_json"),
        Encoders.STRING());
    ds.show(5, false);
    ds.printSchema();

    Dataset<Row> dfJson = spark.read().json(ds);
    dfJson.show(5, false);
    dfJson.printSchema();

    dfJson = dfJson
        .withColumn("emp", explode(dfJson.col("employee")))
        .drop("employee");
    dfJson.show(5, false);
    dfJson.printSchema();

    dfJson = dfJson
        .withColumn("emp_name",
            concat(
                dfJson.col("emp.name.firstName"),
                lit(" "),
                dfJson.col("emp.name.lastName")))
        .withColumn("emp_address",
            concat(dfJson.col("emp.address.street"),
                lit(" "),
                dfJson.col("emp.address.unit")))
        .withColumn("emp_city", dfJson.col("emp.address.city"))
        .drop("emp");
    dfJson.show(5, false);
    dfJson.printSchema();
  }
}

File: src/main/python/lab951_csv_with_embdedded_json_auto/csvWithEmbdeddedJsonAutomaticJsonifierApp.py
"""
 Ingesting a CSV with embedded JSON.

 @author rambabu.posa
"""
import os
from pyspark.sql import (SparkSession, DataFrame, functions as F)

def get_absolute_file_path(path, filename):
    current_dir = os.path.dirname(__file__)
    relative_path = "{}{}".format(path, filename)
    absolute_file_path = os.path.join(current_dir, relative_path)
    return absolute_file_path

def jsonifier(df):
    print(df.schema)
    return ''


def main(spark, path, filename):

    absolute_file_path = get_absolute_file_path(path, filename)

    df = spark.read.option("header", True) \
        .option("delimiter", "|") \
        .option("inferSchema", True) \
        .csv(absolute_file_path)
    return df

'''
  It works only for this scenario
  To write this as a generic function, please reimplement this using recursion technique 
'''
def flatten_nested_structure(df:DataFrame) -> DataFrame:
    array_cols = [c[0] for c in df.dtypes if c[1][:5] == 'array']
    new_df = df.select('*', F.explode(*array_cols)).drop('employee')

    struct_cols = [c[0] for c in new_df.dtypes if c[1][:6] == 'struct']
    new_df2 = new_df.select(*[c + ".*" for c in struct_cols])

    struct_cols2 = [c[0] for c in new_df2.dtypes if c[1][:6] == 'struct']
    new_df3 = new_df2.select(*[c + ".*" for c in struct_cols2])
    return new_df3

def rename_all_df_columns(df,new_column_names):
    return df.toDF(*new_column_names)

if __name__ == '__main__':
    # Creates a session on a local master
    spark = SparkSession.builder.appName("Processing of invoices") \
        .master("local[*]") \
        .getOrCreate()
    spark.sparkContext.setLogLevel('warn')

    path = '../../../../data/misc/'
    filename = 'csv_with_embedded_json2.csv'

    df = main(spark, path, filename)
    df.show(5, False)
    df.printSchema()

    new_df = spark.read.json(df.rdd.map(lambda r: r.emp_json))
    new_df.show(5, False)
    new_df.printSchema()

    flat_df = flatten_nested_structure(new_df)
    flat_df.show(5, False)
    flat_df.printSchema()

    flat_df = flat_df.withColumn("emp_name", F.concat(F.col("firstName"), F.lit(" "), F.col("lastName"))) \
        .withColumn("emp_address", F.concat(F.col("street"), F.lit(" "), F.col("unit"))) \
        .withColumn("emp_city", F.col("city")) \
        .drop('firstName', 'lastName', 'street', 'unit', 'city')

    flat_df.show()
    flat_df.printSchema()

    trimmed_df = df.select('budget', 'dept', 'deptId', 'location') \
        .withColumn("idx", F.monotonically_increasing_id())
    flat_df = flat_df.withColumn("idx", F.monotonically_increasing_id())

    final_df = trimmed_df.join(flat_df, 'idx', 'outer')
    final_df.show()
    final_df.printSchema()

    spark.stop()